src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -0,0 +1,1413 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Management.Automation.Language;
+using System.Management.Automation.Runspaces;
+using System.Reflection;
+using System.Text.RegularExpressions;
+using Microsoft.PowerShell.Commands;
+
+using CimClass = Microsoft.Management.Infrastructure.CimClass;
+using CimInstance = Microsoft.Management.Infrastructure.CimInstance;
+using CimPropertyDeclaration = Microsoft.Management.Infrastructure.CimPropertyDeclaration;
+
+namespace System.Management.Automation
+{
+    /// <summary>
+    /// Enum describing permissions to use runtime evaluation during type inference
+    /// </summary>
+    [Flags]
+    public enum TypeInferenceRuntimePermissions {
```

Comment:
The goal here is to end up with a public type inference API.  Script analyzer is one obvious tool that would benefit (in fact they have their own implementation, and a shared implementation would be better I think.)